### PR TITLE
Add support for Mixtral8x7B in TT-Transformers

### PR DESCRIPTION
**Pull request by tt-hzhou**
### Problem description
We want to add support to Mixtral8x7B in the tt-transformer library while deprecating the old version in the demo folder.

### What's changed
Module specific to Mixtral are added as separate class in tt-transformer while maintaining compatibility with the general purpose decoder and full model code. 
Individual test cases were created to test performance and accuracy of individual module and mixtral as part of tt-transformer.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/17479547060/job/49647342355

- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
https://github.com/tenstorrent/tt-metal/actions/runs/17587550643/job/49961008895

- [x] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
https://github.com/tenstorrent/tt-metal/actions/runs/17587576388/job/49961114567

- [x] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work 
https://github.com/tenstorrent/tt-metal/actions/runs/17587589748/job/49961043220

- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
https://github.com/tenstorrent/tt-metal/actions/runs/17736256408

- [x] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
https://github.com/tenstorrent/tt-metal/actions/runs/17763560324/job/50482966111

- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/model_config.py [hunk 1]
```diff
@@ -1813,9 +1851,108 @@ def load_state_dict(self):
...
+            self.initialize_mixture_of_experts_configs()
+            self.moe = True
+            self.num_experts = 8
         return state_dict
 
+    def initialize_mixture_of_experts_configs(self):
+        self.model_config[
+            "DECODE_MIXTRAL_MLP_W1_PRG_CONFIG"
+        ] = ttnn.MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig(
+            in0_block_w=2,  # 14336 / 32 / 8 / 8 the largest divisor
```
Comment (yieldthought) [inline models/tt_transformers/tt/model_config.py]: Please calculate the value e.g. hidden_dim / tile_size / num_devices / whatever. In TTT we do not include lovingly hand-crafted integers that will only work for the exact dimensions and chip counts specified.

I guess these were hard-coded in the mixtral codebase. TTT provides helper functions to determine valid dram sharded matmul program configs - please try swapping these out for those.

File: models/tt_transformers/tests/mixtral/tale-of-two-cities.txt.bz2
Comment (yieldthought) [inline models/tt_transformers/tests/mixtral/tale-of-two-cities.txt.bz2:1]: I don't want multiple copies of the tale of two cities in the tt-metal codebase! Please remove this and update the code that uses it to point to models/tt_transformers/tests/tale-of-two-cities.txt.bz2 instead.

File: models/tt_transformers/tt/decoder.py [hunk 2]
```diff
@@ -80,9 +104,12 @@ def __init__(
                 weight_dtype=ttnn.bfloat16,
                 weight_key="attention_norm",
                 is_distributed=self.args.is_distributed_norm,
-                add_unit_offset=self.args.rms_norm_add_unit_offset,
-                sharded_program_config=self.model_config["SHARDED_NORM_ATTN_PRGM_CFG"],
-                sharded_output_config=self.model_config["SHARDED_ATTN_INPUT_MEMCFG"],
+                sharded_program_config=self.model_config[
+                    "SHARDED_NORM_ATTN_PRGM_CFG"
+                ],  # LayerNormShardedMultiCoreProgramConfig(compute_with_storage_grid_size=(x=8,y=4),subblock_w=4,block_h=1,block_w=4,inplace=0)
```
Comment (yieldthought) [inline models/tt_transformers/tt/decoder.py]: Please remove these comments

File: models/tt_transformers/tt/decoder.py [hunk 3]
```diff
@@ -80,9 +104,12 @@ def __init__(
...
+                sharded_output_config=self.model_config[
+                    "SHARDED_ATTN_INPUT_MEMCFG"
+                ],  # MemoryConfig(memory_layout=TensorMemoryLayout::WIDTH_SHARDED,buffer_type=BufferType::L1,shard_spec=ShardSpec(grid={[(x=0,y=0) - (x=7,y=3)]},shape={32, 128},orientation=ShardOrientation::ROW_MAJOR,mode=ShardMode::PHYSICAL,physical_shard_shape=std::nullopt))
```
Comment (yieldthought) [inline models/tt_transformers/tt/decoder.py]: See above

File: models/tt_transformers/tt/decoder.py [hunk 4]
```diff
@@ -101,9 +128,12 @@ def __init__(
...
                 is_distributed=self.args.is_distributed_norm,
-                add_unit_offset=self.args.rms_norm_add_unit_offset,
-                sharded_program_config=self.model_config["SHARDED_NORM_MLP_PRGM_CFG"],
-                sharded_output_config=self.model_config["SHARDED_MLP_INPUT_MEMCFG"],
+                sharded_program_config=self.model_config[
+                    "SHARDED_NORM_MLP_PRGM_CFG"
+                ],  # LayerNormShardedMultiCoreProgramConfig(compute_with_storage_grid_size=(x=8,y=1),subblock_w=4,block_h=1,block_w=16,inplace=0)
+                sharded_output_config=self.model_config[
+                    "SHARDED_MLP_INPUT_MEMCFG"
+                ],  # MemoryConfig(memory_layout=TensorMemoryLayout::WIDTH_SHARDED,buffer_type=BufferType::L1,shard_spec=ShardSpec(grid={[(x=0,y=0) - (x=7,y=0)]},shape={32, 512},orientation=ShardOrientation::ROW_MAJOR,mode=ShardMode::PHYSICAL,physical_shard_shape=std::nullopt))
```
Comment (yieldthought) [inline models/tt_transformers/tt/decoder.py]: All of these comments should be removed

File: models/tt_transformers/tt/decoder.py [hunk 5]
```diff
@@ -80,9 +104,12 @@ def __init__(
                 weight_dtype=ttnn.bfloat16,
                 weight_key="attention_norm",
                 is_distributed=self.args.is_distributed_norm,
-                add_unit_offset=self.args.rms_norm_add_unit_offset,
```
Comment (yieldthought) [inline models/tt_transformers/tt/decoder.py]: Removing this seems like an odd part of this PR. Was it done for a good reason or by accident?

File: models/tt_transformers/tt/decoder.py [hunk 6]
```diff
@@ -101,9 +128,12 @@ def __init__(
                 weight_dtype=ttnn.bfloat16,
                 weight_key="ffn_norm",
                 is_distributed=self.args.is_distributed_norm,
-                add_unit_offset=self.args.rms_norm_add_unit_offset,
```
Comment (yieldthought) [inline models/tt_transformers/tt/decoder.py]: Was removing add_unit_offset intentional?

File: models/tt_transformers/tt/model_config.py [hunk 7]
```diff
@@ -1813,9 +1851,108 @@ def load_state_dict(self):
         for k in keys_dict:
             if any([r in k for r in remv]):
                 state_dict.pop(k)
-
+        if self.is_mixture_of_experts:
+            self.initialize_mixture_of_experts_configs()
+            self.moe = True
+            self.num_experts = 8
```
Comment (yieldthought) [inline models/tt_transformers/tt/model_config.py]: Can we get the number of experts from the huggingface config definition?

File: models/tt_transformers/tt/model_config.py [hunk 8]
```diff
@@ -1826,6 +1963,16 @@ def create_dram_sharded_mem_config(self, k, n):
         )
         return ttnn.MemoryConfig(ttnn.TensorMemoryLayout.WIDTH_SHARDED, ttnn.BufferType.DRAM, shard_spec)
 
+    def _create_dram_sharded_mem_config(self, k, n):
```
Comment (yieldthought) [inline models/tt_transformers/tt/model_config.py]: Uhhh is this a duplicate of create_dram_sharded_mem_config?

File: tests/scripts/t3000/run_t3000_demo_tests.sh [hunk 9]
```diff
@@ -230,11 +230,14 @@ run_t3000_mixtral_tests() {
...
+  echo "LOG_METAL: Running run_t3000_tt-transformer_mixtral8x7b_tests"
 
   # mixtral8x7b 8 chip demo test - 100 token generation with general weights (env flags set inside the test)
-  pytest -n auto models/demos/t3000/mixtral8x7b/demo/demo.py --timeout=720 ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/demo/demo_with_prefill.py --timeout=720 ; fail+=$?
+  # pytest -n auto models/demos/t3000/mixtral8x7b/demo/demo.py --timeout=720 ; fail+=$?
+  # pytest -n auto models/demos/t3000/mixtral8x7b/demo/demo_with_prefill.py --timeout=720 ; fail+=$?
+  mixtral8x7=/mnt/MLPerf/huggingface/hub/models--mistralai--Mixtral-8x7B-v0.1/snapshots/fc7ac94680e38d7348cfa806e51218e6273104b0
+
+  CI=true HF_MODEL=$mixtral8x7 pytest -n auto models/tt_transformers/demo/simple_text_demo.py -k "not performance-ci-stress-1" --timeout=20800 ; fail+=$?
```
Comment (yieldthought) [inline tests/scripts/t3000/run_t3000_demo_tests.sh]: That's one hell of a timeout, infra will want that tightened up to ~2x the expected runtime or similar.

File: tests/scripts/t3000/run_t3000_unit_tests.sh [hunk 10]
```diff
@@ -412,15 +412,27 @@ run_t3000_mixtral_tests() {
...
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_attention.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_mlp.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_rms_norm.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_embedding.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_moe.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_decoder.py ; fail+=$?
-  # Mixtral prefill tests
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_mlp_prefill.py ; fail+=$?
-  pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_moe_prefill.py ; fail+=$?
+  # pytest -n auto models/demos/t3000/mixtral8x7b/tests/test_mixtral_attention.py ; fail+=$?
```
Comment (yieldthought) [inline tests/scripts/t3000/run_t3000_unit_tests.sh]: Remove these rather than leaving comments in the scripts.

File: .github/workflows/t3000-demo-tests-impl.yaml [hunk 11]
```diff
@@ -84,7 +84,7 @@ jobs:
             -e TT_METAL_HOME=${{ github.workspace }}
             -e ARCH_NAME=${{ matrix.test-group.arch }}
             -e LD_LIBRARY_PATH=${{ github.workspace }}/build/lib
-            -v /mnt/MLPerf:/mnt/MLPerf:ro
+            -v /mnt/MLPerf:/mnt/MLPerf:rw
```
Comment (yieldthought) [inline .github/workflows/t3000-demo-tests-impl.yaml]: Remember to put this back after generating the cache files

**Review by yieldthought — CHANGES_REQUESTED** [review]

**Review by yieldthought — APPROVED** [review]
Approved but remember to make the CI filesystem ro again and remove commented out lines!

File: .github/workflows/t3000-demo-tests-impl.yaml [hunk 12]
```diff
@@ -23,6 +23,7 @@ jobs:
       fail-fast: false
       matrix:
         test-group: [
+          { name: "t3k_mixtral_tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 50, owner_id: U03PUAKE719}, # Miguel Tairum
```
Comment (skhorasganiTT) [inline .github/workflows/t3000-demo-tests-impl.yaml]: You should be the owner here right?

File: .github/workflows/t3000-unit-tests-impl.yaml [hunk 13]
```diff
@@ -36,7 +36,7 @@ jobs:
           { name: "t3k llama3.2-90b-vision tests", arch: wormhole_b0, cmd: run_t3000_llama3.2-90b-vision_unit_tests, timeout: 30, label: pipeline-functional, owner_id: U07RY6B5FLJ},  #Gongyu Wang
           #{ name: "t3k llama3.1-70b tests", arch: wormhole_b0, cmd: run_t3000_llama3.1-70b_tests, timeout: 30, owner_id: U03PUAKE719},  #Miguel Tairum Cruz
           { name: "t3k llama3.2-90b tests", arch: wormhole_b0, cmd: run_t3000_llama3.2-90b_tests, timeout: 30, label: pipeline-functional, owner_id: U07RY6B5FLJ},  #Gongyu Wang
-          #{ name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
+          { name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
```
Comment (skhorasganiTT) [inline .github/workflows/t3000-unit-tests-impl.yaml]: You should be the owner here right?

File: log-tmp.txt [hunk 14]
```diff
@@ -0,0 +1,536 @@
+Creating virtual env in: /localdev/hzhou/tt-metal/python_env
+Forcefully using a version of pip that will work with our view of editable installs
+Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cpu
```
Comment (skhorasganiTT) [inline log-tmp.txt]: Please delete this file

File: models/tt_transformers/tt/mixtral_mlp.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/mixtral_mlp.py]: Should be 2025, same comment for other new files

File: models/tt_transformers/tt/model_config.py [hunk 16]
```diff
@@ -471,6 +475,8 @@ def __init__(
         self.dummy_weights = dummy_weights
         self.cache_hf_flag = cache_hf  # Whether to cache HF model to avoid multiple loads (uses extra memory)
         self.cached_hf_model = None  # Save any HF model object to avoid loading it multiple times for reference methods
+        self.num_experts_per_tok = 2
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model_config.py]: Possible to get from hf config?

File: models/tt_transformers/tt/model_config.py [hunk 17]
```diff
@@ -705,6 +713,17 @@ def __init__(
                 packer_l1_acc=False,
             )
 
+            # Configure data precision and math fidelity for tensors and kernels
+            self.model_config["COMPUTE_KERNEL_CONFIG_HIFI2"] = self.compute_kernel_config_hifi2
+            # Mixtral
+            self.model_config["PREFILL_MLP_COMPUTE_CONFIG"] = self.compute_kernel_config_lofi
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model_config.py]: Maybe should prefix with MIXTRAL since only used by that model?

File: models/tt_transformers/tt/model_config.py [hunk 18]
```diff
@@ -705,6 +713,17 @@ def __init__(
...
+            self.model_config["GATE_MM_OUTPUT_KERNEL_CONFIG"] = self.compute_kernel_config_lofi
+            # end mixtral
+
+            if self.optimizations is None:
+                self.optimizations = DecodersPrecision.accuracy(num_decoders=self.n_layers, model_name=self.model_name)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model_config.py]: Should this be defaulting to accuracy? cc @yieldthought

File: models/tt_transformers/tt/decoder.py [hunk 19]
```diff
@@ -59,16 +61,38 @@ def __init__(
...
-            mesh_device=mesh_device,
-            tt_ccl=self.tt_ccl,
-            args=args,
-            state_dict=state_dict,
-            weight_cache_path=weight_cache_path,
-            layer_num=layer_num,
-            dtype=dtype,
-            model_config=self.model_config,
-        )
+        if self.args.is_mixture_of_experts:
```
Comment (handrewsTT) [inline models/tt_transformers/tt/decoder.py]: @tt-hzhou You could do something like `if hasattr(self, is_mixture_of_experts) and self.args.is_mixture_of_experts:` then you don't need to modify the gemma model_config (or any other temporary model config duplicates we might have in flight)

**Review by skhorasganiTT — APPROVED** [review]

**Review by roseli-TT — APPROVED** [review]
